Added Psalm and PhpStan for static analysis

driehle opened the pull request:
Fixes #247, #248

driehle committed:
added PhpStan and Psalm

driehle committed:
added CI workflow

driehle committed:
fixed Psalm issues

driehle committed:
fixed PhpStan issues

driehle committed:
added missing interfaces to Module class

driehle committed:
updated branch name in readme

driehle committed:
updated composer scripts and `.gitattributes`

SenseException commented on tests/Doctrine/CustomDefaultRepositoryTest.php:
```diff
@@ -16,7 +18,8 @@ public function testCustomDefaultRepository(): void
 
-        self::assertInstanceOf(DefaultDocumentRepository::class, $repository);
-        self::assertTrue($repository->isCustomDefaultDocumentRepository());
+        $this->assertInstanceOf(DefaultDocumentRepository::class, $repository);
+        assert($repository instanceof DefaultDocumentRepository);
+        $this->assertTrue($repository->isCustomDefaultDocumentRepository());
```
Why change `self::` to `$this->` when the rest is using static too?

driehle replied:
Hm, indeed. Is there any of these options preferred over the other? I think I mostly saw `$this` so far.

SenseException replied:
Since they are static, I personally prefer the static writing.

driehle replied:
Since DoctrineModule and DoctrineORMModule use `$this`, I refactored this project to follow these standards.

SenseException replied:
I'm fine with that as long as it's not mixed.

SenseException commented on tests/Doctrine/AnnotationTest.php:
```diff
@@ -11,7 +11,7 @@ class AnnotationTest extends AbstractTest
         self::markTestIncomplete();
-        $documentManager = $this->getDocumentManager();
-        $metadata        = $documentManager->getClassMetadata('DoctrineMongoODMModuleTest\Assets\Document\Annotation');
+        //$documentManager = $this->getDocumentManager();
+        //$metadata = $documentManager->getClassMetadata('DoctrineMongoODMModuleTest\Assets\Document\Annotation');
```
Why keeping the commented code?

driehle replied:
Well, I don't know what this test was supposed to do. It obviously didn't do anything in the last years, because of the `self::markTestIncomplete()` line above.

Since there are no other tests in that file, the whole file is somehow obsolete.

SenseException replied:
I'm actually thinking about removing this test. It seems like it lost it's meaning during development.

driehle replied:
Ok, I removed it with 7e23850

SenseException commented on tests/Doctrine/ConnectionFactoryTest.php:
```diff
@@ -16,8 +16,8 @@ class ConnectionFactoryTest extends AbstractTest
-    /** @var mixed[] $connectionFactory */
-    private $connectionFactory = [];
+    /** @var ConnectionFactory $connectionFactory */
+    private $connectionFactory = null;
```
```suggestion
    private $connectionFactory;
```

The connection-factory is set in `setUp()`

driehle committed:
streamline PhpUnit assertions using non-static calls

driehle committed:
removed old incomplete test case

This test case was introduced in 8f887992687f06f3f2c8ce47c3cf803f4ae5bf99 and was related to testing persistence. However, it doesn't do anything and hasn't done anything for years, so we remove it.
